sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala

```diff
@@ -1029,10 +1029,10 @@ class DataFrame private[sql](
 
     val elementTypes = schema.toAttributes.map { attr => (attr.dataType, attr.nullable) }
     val names = schema.toAttributes.map(_.name)
+    val convert = CatalystTypeConverters.getConverterForType(schema)
```

Comment:
I think you could call `createToCatalystConverter` here, which returns an `Any => Any` function that you can cast to the appropriate converter type.  This prevents us from having to increase the visibility of the internal `CatalystTypeConverter` APIs.